### ISSUE TITLE
p2p: drop misbehaving peer

### ISSUE THREAD

**Issue by countvonzero** [spacemeshos/go-spacemesh#4088]
## Description

when the node receives a invalid message via gossip or sync, the node should drop the peer

specific to sync, we also need to identity when a peer failed to deliver content for hash that it claimed to have.

related to #3551

**Comment by dshulyak** [comment]
#### invalid messages (or any violation of the protocol rules)
any syntactically invalid message can't be sent by mistake. they are also may not be cheap
to download or verify. therefore we want to drop peer immediately. and schedule peer for blacklisting by ip address. blacklisting it immediately is unnecessary risk for network health, as we will do it by ip address it may blacklist other nodes as well.

we can make a threshold such as 5  syntactically invalid messages result in blacklisting peer for 30 minutes.

#### contextually invalid messages
if peer sends contextually invalid message (such as atx equivocation) it can't be disqualified immediately, as it may not know that the identity equivocated. we don't have strict rules in the implementation to know when peer actually misbehaves. when we will add them handler should report rejection.

#### wasting bandwidth (duplicates, overflowing limits)
libp2p gossipsub provides several configuration parameters to set expectations on message limits. peer will be throttled by dropping published messages, it may potentially waste bandwidth. we need to evaluate such ddos separately, as right now we probably have easier means to make node unusable. 

so for mid-term we should setup provided options to penalize peers that overuse limits or send already known messages.

#### integration
both gossipsub and sync should use decorator, that drops a peer on reject and calls book.Penalize(id, ip).
on book.Penalize track penalty per ip, once penalty grows to blacklist threshold - ip is blacklisted for some delay. penalty for the duration of several layers should be sufficient to prevent ddos.

book should not add or share addresses of the peers that are currently blacklisted. 

libp2p needs to reject connections from blacklisted peers, this is to be implemented by implemeting ConnGater interface. that is configured for libp2p.Host.

book must maintain time based cache for blacklisted peers that expires periodically. for simplicity we can run goroutine in background that expires ips periodically. number of blacklisted ips should be limited to some reasonable number, taking into account memory requirement. it is also not necessary to persist them.

**Comment by countvonzero** [comment]
> blacklisting it immediately is unnecessary risk for network health, as we will do it by ip address it may blacklist other nodes as well.

would like to learn in what scenario would blocking one IP address also blacklist other nodes.

**Comment by piersy** [comment]
> would like to learn in what scenario would blocking one IP address also blacklist other nodes.

@countvonzero IP addresses being ephemeral, there is a chance that the IP could be assigned to a different node.

**Comment by dshulyak** [comment]
I am mostly worried about routable ips, they are managed by ISP's and not ephemeral (unless i misunderstood the meaning). Usually if ISP doesn't give everyone public ip address, your address will be masqueraded on ISP router, so you can end up sharing it with random nodes. The same if you use VPN/TOR to hide actual ip address.

**Comment by noamnelke** [comment]
I agree that IP bans should not be permanent. I can think of several scenarios where an IP address is either shared by several users or rotated between users.

The spec looks good, in general.

I agree about priorities:
1. Immediately drop any neighbor that sends a single syntactically invalid message.
2. Track drops per IP address, then blacklist for 30 minutes any IP address that got dropped 5 times in <30 minutes. This way if someone deliberately tries to maximize the rate of invalid messages, they can send us no more than 10 per hour from a single IP (one every 6 minutes + 1 second). We should obviously make these thresholds configurable.
3. Track and punish peers that overuse allowed-but-useless messages (mostly contextually invalid messages).

We should only do 1+2 at this point, as 3 is harder to define and implement and also less of an issue.

### Regarding duplicates:
Sending a message more than once can be treated like sending a syntactically invalid message **if we can technically keep track on both the sending and receiving ends**. A node should never send a message more than once on a link and also never send back a message to the peer that originally sent it. I don't know if it's feasible to implement this, since we may not keep track of what we received from whom. If we do, or it's easy, then I think we should consider duplicates as grounds for dropping a peer.

**Comment by dshulyak** [comment]
> Sending a message more than once can be treated like sending a syntactically invalid message if we can technically keep track on both the sending and receiving ends. A node should never send a message more than once on a link and also never send back a message to the peer that originally sent it. I don't know if it's feasible to implement this, since we may not keep track of what we received from whom. If we do, or it's easy, then I think we should consider duplicates as grounds for dropping a peer.

there is a time-based cache inside the libp2p router to track that. if timings in caches are not perfectly aligned there will be small amount of duplicates from otherwise honest nodes. it would be too risky to disconnect based on information in this cache

**Comment by countvonzero** [comment]
moving from Done->Backlog as the syncing part is not done.

**Comment by dshulyak** [comment]
yes it is not done. but what do you mean by syncing part? peer will be dropped too if syncer rejects the message

**Comment by countvonzero** [comment]
> yes it is not done. but what do you mean by syncing part? peer will be dropped too if syncer rejects the message

it's done on the validation path yes. but not at a higher level. for example, if a peer tells a node that layer data contains hashes A/B/C, but cannot produce the content, that's also a malicious peer and should be blacklisted.

but i see now the the original description describes a narrower scope of the issue title, hence the confusion.
